subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/LazyTaskExecution.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedMap;
+import com.google.common.collect.ImmutableSortedSet;
+import org.gradle.api.internal.OverlappingOutputs;
+import org.gradle.internal.id.UniqueId;
+
+import java.util.Map;
+
+public class LazyTaskExecution extends TaskExecution {
+    private ImmutableSortedMap<String, Long> inputFilesSnapshotIds;
+    private ImmutableSortedMap<String, Long> outputFilesSnapshotIds;
+    private Long discoveredFilesSnapshotId;
+    private final FileSnapshotRepository snapshotRepository;
+    private ImmutableSortedMap<String, FileCollectionSnapshot> inputFilesSnapshot;
+    private ImmutableSortedMap<String, FileCollectionSnapshot> outputFilesSnapshot;
+    private FileCollectionSnapshot discoveredFilesSnapshot;
+
+    /**
+     * Creates a mutable copy of the given snapshot.
+     */
+    public LazyTaskExecution(FileSnapshotRepository snapshotRepository, TaskExecutionSnapshot taskExecutionSnapshot) {
+        this(
+            snapshotRepository,
+            taskExecutionSnapshot.getBuildInvocationId(),
+            taskExecutionSnapshot.getTaskImplementation(),
+            taskExecutionSnapshot.getTaskActionsImplementations(),
+            taskExecutionSnapshot.getInputProperties(),
+            taskExecutionSnapshot.getCacheableOutputProperties(),
+            taskExecutionSnapshot.getDeclaredOutputFilePaths(),
+            taskExecutionSnapshot.isSuccessful(),
+            null
```

Comment:
Does it make sense to store the detected overlapping outputs in the task history? I currently didn't come up with a use case, but perhaps you have some ideas.